flycheck.el

```diff
@@ -8831,32 +8838,71 @@ Relative paths are relative to the file being checked."
                  (flycheck-error-message err)))
               errors))
 
+(defun flycheck-rust-valid-crate-type-p (crate-type)
+  "Whether CRATE-TYPE is a valid target type for Cargo.
+
+A valid Cargo target type is one of `lib', `bin', `example',
+`test' or `bench'."
+  (member crate-type '("lib" "bin" "example" "test" "bench")))
+
 (flycheck-define-checker rust-cargo
   "A Rust syntax checker using Cargo.
 
 This syntax checker requires Rust 1.15 or newer.  See URL
 `https://www.rust-lang.org'."
   :command ("cargo" "rustc"
-            (eval (cond
-                   ((string= flycheck-rust-crate-type "lib") "--lib")
-                   (flycheck-rust-binary-name
-                    (list "--bin" flycheck-rust-binary-name))))
+            (eval (concat "--" flycheck-rust-crate-type))
+            ;; All crate targets except "lib" need a binary name
+            (eval (unless (string= flycheck-rust-crate-type "lib")
+                    flycheck-rust-binary-name))
             "--message-format=json"
             (eval flycheck-cargo-rustc-args)
             "--" "-Z" "no-trans"
-            (option-flag "--test" flycheck-rust-check-tests)
-            (option-list "-L" flycheck-rust-library-path concat)
+            ;; Passing the "--test" flag when the target is a test binary or
+            ;; bench is unnecessary and triggers an error.
+            (eval (when flycheck-rust-check-tests
+                    (unless (member flycheck-rust-crate-type '("test" "bench"))
+                      "--test")))
             (eval flycheck-rust-args))
   :error-parser flycheck-parse-cargo-rustc
   :error-filter flycheck-rust-error-filter
   :error-explainer flycheck-rust-error-explainer
   :modes rust-mode
-  :predicate (lambda ()
-               ;; Since we build the entire project with cargo rustc we require
-               ;; that the buffer is saved.  And of course, we also need a Cargo
-               ;; file :)
-               (and (flycheck-buffer-saved-p)
-                    (locate-dominating-file (buffer-file-name) "Cargo.toml"))))
+  :predicate flycheck-buffer-saved-p
+  :enabled (lambda ()
+             (and (locate-dominating-file (buffer-file-name) "Cargo.toml")
+                  (flycheck-rust-valid-crate-type-p flycheck-rust-crate-type)
```

Comment:
The check to `rust-valid-crate-type`?  Well, `:enabled` is rather recent, so not all checkers use it.  My thinking was that if we enable the checker with an invalid crate type, the user will get a suspicious checker state.  Since we have `:verify`, we might as well disable a checker with an invalid configuration.

Though there's an unforeseen side-effect: disabling the checker will allow for the plain `rust` checker to run on the file.  So I'm not sure anymore if that's a better situation :/